Fix nonlinsolve TypeError when nfloat produces nan

jhaayushkumar opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #28646

#### Brief description of what is fixed or changed

Fixed `TypeError: 'NoneType' object is not iterable` in [nonlinsolve](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/solvers/solveset.py:3920:0-4152:27) when processing equations with float coefficients.

**Root Cause:**
The [_handle_poly](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/solvers/solveset.py:3806:0-3917:29) function was converting polynomial coefficients back to float using `nfloat()`, which produced `nan` for certain complex expressions. This caused [solve_poly_system](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/solvers/polysys.py:38:0-97:51) to return `None`, leading to a crash when attempting to iterate over the result.

**before**
On Running:- 
```python
lst=[x**2*(y-0.508793242825092)**2-sin(pi/8)*sqrt(0.508793242825092**2), 
    (x-1)**2*y**2-0.508793242825092**2]
nonlinsolve(lst, x,y)
  ```

**Output:**
```python
TypeError: 'NoneType' object is not iterable
```
after fix:- 
```python
lst=[x**2*(y-0.508793242825092)**2-sin(pi/8)*sqrt(0.508793242825092**2), 
    (x-1)**2*y**2-0.508793242825092**2]
nonlinsolve(lst, x,y)

FiniteSet(
    (solution_x1, solution_y1),
    (solution_x2, solution_y2),
    ...
    (Total 8 set)
)
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, [log(-x)](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/solvers/solveset.py:1971:0-2026:13) incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
  * Fixed a bug where nonlinsolve would fail with an exception for some systems of polynomial equations that mix floats and surds.

<!-- END RELEASE NOTES -->

oscarbenjamin commented:
Checking for nan is the wrong approach. The expression that numerically evaluates to nan should not be produced in the first place.

jhaayushkumar commented on sympy/solvers/solveset.py:
```diff
@@ -3874,8 +3874,12 @@ def _handle_poly(polys, symbols):
+                poly_eqs = list(basis)
+            else:
+                poly_sol = [dict(zip(symbols, res)) for res in result]
+                poly_eqs = no_equations
```
> Why is this changed?
> 
> Under what conditions does `solve_poly_system` return `None`?

I added it while debugging. solve_poly_system returns None when 
basis is just [1]. But your fix prevents that, so removing it.

jhaayushkumar commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -2030,6 +2030,17 @@ def test_nonlinsolve_inexact():
+    ]
+    result = nonlinsolve(lst, x, y)
+    assert isinstance(result, FiniteSet)
+    assert len(result) > 0
```
I changed tests by using hardcoded value in sol array and then perform this
```python
assert all(abs(res.args[i][j] - sol[i][j]) < 1e-9
               for i in range(8) for j in range(2))
```
Did this ok or any changes required

oscarbenjamin commented:
Can you squash the commits and improve the release note.

jhaayushkumar committed:
Fix nonlinsolve TypeError from polynomial domain confusion

Fixes #28646

Use .evalf() before nsimplify() to normalize polynomial domains,
preventing RR/EX confusion in Groebner basis computation.

oscarbenjamin commented:
Looks good.
